File: DSA/SAMAGRA_590014891/Day_93_Question_1.c
#include <stdio.h>

int houseRobber(int houses[], int n) {
    if (n == 0) return 0;
    if (n == 1) return houses[0];

    int dp[n];
    dp[0] = houses[0];
    dp[1] = (houses[0] > houses[1]) ? houses[0] : houses[1];

    for (int i = 2; i < n; i++) {
        int rob = houses[i] + dp[i-2];
        int skip = dp[i-1];
        dp[i] = (rob > skip) ? rob : skip;
    }
    return dp[n-1];
}

int main() {
    int houses1[] = {1, 2, 3, 1};
    int houses2[] = {2, 7, 9, 3, 1};

    printf("Max loot (case 1): %d\n", houseRobber(houses1, 4));
    printf("Max loot (case 2): %d\n", houseRobber(houses2, 5));

    return 0;
}

File: DSA/VenikaPanwar_590013755/Day_93_Question_1.py
def rob(nums):
    """
    Finds maximum amount that can be robbed without robbing adjacent houses.
    Time: O(n), Space: O(n)
    """
    if not nums:
        return 0
    if len(nums) == 1:
        return nums[0]
    
    n = len(nums)
    dp = [0] * n
    dp[0] = nums[0]
    dp[1] = max(nums[0], nums[1])
    
    for i in range(2, n):
        # Either rob current house + best from i-2, or skip current house
        dp[i] = max(dp[i-1], dp[i-2] + nums[i])
    
    return dp[n-1]

def rob_optimized(nums):
    """
    Space-optimized version using only two variables.
    Time: O(n), Space: O(1)
    """
    if not nums:
        return 0
    if len(nums) == 1:
        return nums[0]
    
    prev2 = nums[0]  # dp[i-2]
    prev1 = max(nums[0], nums[1])  # dp[i-1]
    
    for i in range(2, len(nums)):
        current = max(prev1, prev2 + nums[i])
        prev2, prev1 = prev1, current
    
    return prev1

if __name__ == "__main__":
    # Test cases
    houses1 = [1, 2, 3, 1]
    print(f"Max rob amount for {houses1}: {rob(houses1)}")  # Output: 4
    
    houses2 = [2, 7, 9, 3, 1]
    print(f"Max rob amount for {houses2}: {rob(houses2)}")  # Output: 12
    
    # Test optimized version
    print(f"Max rob amount (optimized) for {houses1}: {rob_optimized(houses1)}")  # Output: 4
    print(f"Max rob amount (optimized) for {houses2}: {rob_optimized(houses2)}")  # Output: 12